File: leku/src/main/java/com/schibstedspain/leku/geocoder/places/GooglePlacesDataSource.java
package com.schibstedspain.leku.geocoder.places;

import android.annotation.SuppressLint;
import android.location.Address;
import android.support.annotation.NonNull;
import com.google.android.gms.common.data.DataBufferUtils;
import com.google.android.gms.location.places.AutocompletePrediction;
import com.google.android.gms.location.places.AutocompletePredictionBufferResponse;
import com.google.android.gms.location.places.GeoDataClient;
import com.google.android.gms.location.places.Place;
import com.google.android.gms.location.places.PlaceBufferResponse;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.tasks.RuntimeExecutionException;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import io.reactivex.Observable;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class GooglePlacesDataSource {
  private final GeoDataClient geoDataClient;

  public GooglePlacesDataSource(GeoDataClient geoDataClient) {
    this.geoDataClient = geoDataClient;
  }

  public Observable<List<Address>> getFromLocationName(String query, LatLngBounds latLngBounds) {
    return Observable.defer(() -> {
      Task<AutocompletePredictionBufferResponse> results =
          geoDataClient.getAutocompletePredictions(query, latLngBounds, null);
      try {
        Tasks.await(results, 6, TimeUnit.SECONDS);
      } catch (ExecutionException | InterruptedException | TimeoutException ignored) {
      }

      try {
        AutocompletePredictionBufferResponse autocompletePredictions = results.getResult();
        List<AutocompletePrediction> predictionList = DataBufferUtils.freezeAndClose(autocompletePredictions);
        List<Address> addressList = getAddressListFromPrediction(predictionList);
        return Observable.just(addressList);
      } catch (RuntimeExecutionException e) {
        return Observable.empty();
      }
    });
  }

  @NonNull
  private List<Address> getAddressListFromPrediction(List<AutocompletePrediction> predictionList) {
    List<Address> addressList = new ArrayList<>();
    for (AutocompletePrediction prediction : predictionList) {
      Task<PlaceBufferResponse> placeBufferResponseTask = geoDataClient.getPlaceById(prediction.getPlaceId());
      try {
        Tasks.await(placeBufferResponseTask, 3, TimeUnit.SECONDS);
      } catch (ExecutionException | InterruptedException | TimeoutException ignored) {
      }
      PlaceBufferResponse placeBufferResponse = placeBufferResponseTask.getResult();
      @SuppressLint("RestrictedApi") Place place = placeBufferResponse.get(0);
      addressList.add(mapPlaceToAddress(place));
    }
    return addressList;
  }

  @NonNull
  private Address mapPlaceToAddress(Place place) {
    Address address = new Address(Locale.getDefault());
    address.setLatitude(place.getLatLng().latitude);
    address.setLongitude(place.getLatLng().longitude);
    String addressName = place.getName().toString() + " - " + place.getAddress().toString();
    address.setAddressLine(0, addressName);
    address.setFeatureName(addressName);
    return address;
  }
}

File: leku/src/main/java/com/schibstedspain/leku/geocoder/places/GooglePlacesDataSource.kt
package com.schibstedspain.leku.geocoder.places

import android.location.Address
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.tasks.RuntimeExecutionException
import com.google.android.gms.tasks.Tasks
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.model.RectangularBounds
import com.google.android.libraries.places.api.net.FetchPlaceRequest
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsResponse
import com.google.android.libraries.places.api.net.PlacesClient
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Observable.defer
import java.util.*
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException


private const val PREDICTIONS_WAITING_TIME: Long = 6
private const val PLACE_BY_ID_WAITING_TIME: Long = 3

class GooglePlacesDataSource(private val geoDataClient: PlacesClient) {

    fun getFromLocationName(query: String, latLngBounds: LatLngBounds): Observable<List<Address>> {
        val locationBias = RectangularBounds.newInstance(
            latLngBounds.southwest,
            latLngBounds.northeast)
        return defer {
            val findAutocompletePredictionsRequest = FindAutocompletePredictionsRequest
                .builder()
                .setQuery(query)
                .setLocationBias(locationBias)
                .build()
            val results = geoDataClient.findAutocompletePredictions(findAutocompletePredictionsRequest)
            try {
                Tasks.await(results, PREDICTIONS_WAITING_TIME, TimeUnit.SECONDS)
            } catch (ignored: ExecutionException) {
            } catch (ignored: InterruptedException) {
            } catch (ignored: TimeoutException) {
            }

            try {
                val addressList = getAddressListFromPrediction(results.result)
                return@defer Observable.just(addressList)
            } catch (e: RuntimeExecutionException) {
                return@defer Observable.just(ArrayList<Address>())
            }
        }
    }

    private fun getAddressListFromPrediction(result: FindAutocompletePredictionsResponse?): List<Address> {
        val addressList = ArrayList<Address>()
        result?.let { result ->
            for (prediction in result.autocompletePredictions) {
                val placeFields = listOf(Place.Field.ID, Place.Field.NAME)
                val fetchPlaceRequest = FetchPlaceRequest.builder(prediction.placeId, placeFields).build()
                val placeBufferResponseTask = geoDataClient.fetchPlace(fetchPlaceRequest)
                try {
                    Tasks.await(placeBufferResponseTask, PLACE_BY_ID_WAITING_TIME, TimeUnit.SECONDS)
                } catch (ignored: ExecutionException) {
                } catch (ignored: InterruptedException) {
                } catch (ignored: TimeoutException) {
                }

                val placeBufferResponse = placeBufferResponseTask.result
                val place = placeBufferResponse?.place
                place?.let {
                    addressList.add(mapPlaceToAddress(it))
                }
            }
        }
        return addressList
    }

    private fun mapPlaceToAddress(place: Place): Address {
        val address = Address(Locale.getDefault())
        place.latLng?.let {
            address.latitude = it.latitude
            address.longitude = it.longitude
        }
        val addressName = place.name.toString() + " - " + place.address.toString()
        address.setAddressLine(0, addressName)
        address.featureName = addressName
        return address
    }
}
